onnxruntime/core/providers/cuda/math/unary_elementwise_ops_impl.cu

```diff
@@ -115,50 +119,139 @@ struct OP_Cast {
   }
 };
 
+#define IMPL_CAST_IMPL(InT, OutT) \
+  void Explicit_Impl_Cast(cudaStream_t stream, const InT* input_data, OutT* output_data, size_t count) { \
+    UnaryElementWiseImpl(stream, input_data, output_data, OP_Cast<InT, OutT>(), count); \
+  }
+
+#define IMPL_CAST_IMPL_THROW(InT, OutT) \
+  void Explicit_Impl_Cast(cudaStream_t stream, const InT* input_data, OutT* output_data, size_t count) { \
+    ORT_THROW("Cast from " #InT " to " #OutT " must define saturate."); \
+  }
+
+
+#define IMPL_CAST_IMPL_FROM(T)      \
+  IMPL_CAST_IMPL(T, half)     \
+  IMPL_CAST_IMPL(T, float)    \
+  IMPL_CAST_IMPL(T, double)   \
+  IMPL_CAST_IMPL(T, int8_t)   \
+  IMPL_CAST_IMPL(T, int16_t)  \
+  IMPL_CAST_IMPL(T, int32_t)  \
+  IMPL_CAST_IMPL(T, int64_t)  \
+  IMPL_CAST_IMPL(T, uint8_t)  \
+  IMPL_CAST_IMPL(T, uint16_t) \
+  IMPL_CAST_IMPL(T, uint32_t) \
+  IMPL_CAST_IMPL(T, uint64_t) \
+  IMPL_CAST_IMPL(T, bool)     \
+  IMPL_CAST_IMPL(T, BFloat16) \
+  IMPL_CAST_IMPL_THROW(T, Float8E4M3FN) \
+  IMPL_CAST_IMPL_THROW(T, Float8E5M2) \
+  IMPL_CAST_IMPL_THROW(T, Float8E4M3FNUZ) \
```

Comment:
Will we ever need this? I understand this will add completeness but wondering if this will ever be used in practice.

Reply:
same applies to the one below.

Reply:
I followed the same design but conversion from/to float8/float16/float are the only ones we need.

Reply:
I got this error: `C++ exception with description "/home/xadupre/github/onnxruntime/onnxruntime/core/session/provider_bridge_ort.cc:1179 onnxruntime::Provider& onnxruntime::ProviderLibrary::Get() [ONNXRuntimeError] : 1 : FAIL : Failed to load library libonnxruntime_providers_cuda.so with error: /home/xadupre/github/onnxruntime/build/linux_cuda/Release/libonnxruntime_providers_cuda.so: undefined symbol: _ZN11onnxruntime4cuda18Explicit_Impl_CastEP11CUstream_stPKdPNS_10Float8E5M2Em
1:`. It needs more work.

Reply:
I suggest keeping them for the time being and fixing this issue in a second PR.